compiler/src/yul/ast_builder.rs

```diff
@@ -0,0 +1,307 @@
+use vyper_parser::ast as vyp;
+use yultsur::yul;
+
+pub type Shared<T> = Rc<RefCell<T>>;
+
+#[allow(dead_code)]
+pub struct ModuleScope<'a> {
+    type_defs: HashMap<&'a str, &'a vyp::TypeDesc<'a>>,
+}
+
+#[allow(dead_code)]
+pub struct ContractScope<'a> {
+    parent: Shared<ModuleScope<'a>>,
+}
+
+#[allow(dead_code)]
+pub struct FunctionScope<'a> {
+    parent: Shared<ContractScope<'a>>,
+}
+
+#[allow(dead_code)]
+pub enum Scope<'a> {
+    Module(Shared<ModuleScope<'a>>),
+    Contract(Shared<ContractScope<'a>>),
+    Function(Shared<FunctionScope<'a>>),
+}
+
+impl<'a> ModuleScope<'a> {
+    fn new() -> Shared<Self> {
+        Rc::new(RefCell::new(ModuleScope {
+            type_defs: HashMap::new(),
+        }))
+    }
+}
+
+impl<'a> ContractScope<'a> {
+    fn new(parent: Shared<ModuleScope<'a>>) -> Shared<Self> {
+        Rc::new(RefCell::new(ContractScope { parent }))
+    }
+}
+
+impl<'a> FunctionScope<'a> {
+    fn new(parent: Shared<ContractScope<'a>>) -> Shared<Self> {
+        Rc::new(RefCell::new(FunctionScope { parent }))
+    }
+}
+
+/// Builds a vector of Yul contracts from a Vyper module.
+pub fn module<'a>(module: &'a vyp::Module<'a>) -> Result<Vec<yul::Object>, CompileError> {
+    let scope = ModuleScope::new();
+
+    Ok(module
+        .body
+        .iter()
+        .map(|stmt| module_stmt(Rc::clone(&scope), &stmt.node))
+        .collect::<Result<Vec<Option<yul::Statement>>, CompileError>>()?
+        .into_iter()
+        .filter_map(|statement| {
+            if let Some(yul::Statement::Object(object)) = statement {
+                return Some(object);
+            }
+
+            None
+        })
+        .collect::<Vec<yul::Object>>())
+}
+
+/// Builds a Yul statement from a Vyper module statement.
+pub fn module_stmt<'a>(
+    scope: Shared<ModuleScope<'a>>,
+    stmt: &'a vyp::ModuleStmt<'a>,
+) -> Result<Option<yul::Statement>, CompileError> {
+    match stmt {
+        vyp::ModuleStmt::TypeDef { .. } => {
+            type_def(scope, stmt)?;
+            Ok(None)
+        }
+        vyp::ModuleStmt::ContractDef { .. } => {
+            contract_def(scope, stmt).map(|object| Some(yul::Statement::Object(object)))
+        }
+        _ => Err(CompileError::static_str(
+            "Unable to translate module statement.",
+        )),
+    }
+}
+
+/// Builds a Yul object from a Vyper contract.
+pub fn contract_def<'a>(
+    scope: Shared<ModuleScope<'a>>,
+    stmt: &'a vyp::ModuleStmt<'a>,
+) -> Result<yul::Object, CompileError> {
+    let new_scope = ContractScope::new(scope);
+
+    if let vyp::ModuleStmt::ContractDef { name: _, body } = stmt {
+        let mut statements = body
+            .iter()
+            .map(|stmt| contract_stmt(Rc::clone(&new_scope), &stmt.node))
+            .collect::<Result<Vec<yul::Statement>, CompileError>>()?;
+
+        // TODO: Use functions from actual contract ABI once the builder is merged.
+        statements.push(yul::Statement::Switch(selectors::switch(vec![])?));
```

Comment:
see comment under `selectors` module about this

```suggestion
        statements.push(yul::Statement::Switch(selectors::dispatch_switch(vec![])?));
```